Improved Google OIDC connector

espadolini opened the pull request:
Optional Google Cloud Identity support to fetch transitive groups (i.e. groups that a user belongs to both directly and indirectly through nested groups) for SSO.

This bumps the version of the `oidc` resources from `v2` to `v3`, supporting both versions, with no change to the actual spec for the resource; the behavior of `v2` connectors will stay the same, while `v3` connectors will not filter the returned groups by domain name, and (depending on the permissions of the configured service account) will attempt to fetch both direct and indirect groups.

Fixes #8122.
Fixes #5521.

espadolini committed:
Optionally fetch transitive groups in the Google OIDC connector

rosstimothy reviewed (commented):
Can you add test coverage? 

Either of the following would allow you to test with out needing to talk to google:
 - Migrating the `groupsFromGsuite*` functions to an interface would allow you to mock the client
 - Utilizing [WithEndpoint](https://pkg.go.dev/google.golang.org/api@v0.65.0/option#WithEndpoint) or [WithGRPCConn](https://pkg.go.dev/google.golang.org/api@v0.65.0/option#WithGRPCConn) would allow you to mock google

espadolini committed:
Refactor the google workspace parts of the OIDC code

espadolini committed:
Further refactoring

This undoes the user account impersonation changes, and always requires
an admin account again.

espadolini committed:
Test coverage

rosstimothy commented on lib/auth/oidc_test.go:
```diff
@@ -176,3 +184,114 @@ func (s *fakeIDP) configurationHandler(w http.ResponseWriter, r *http.Request) {
+		{"bob@foo.example", false, []string{"group1@foo.example"}},
+		{"bob@foo.example", true, []string{"group1@foo.example"}},
+	} {
+		connector.(*types.OIDCConnectorV2).Spec.GoogleTransitiveGroups = testCase.transitive
```
You could create a connector per test case and set this during instantiation to remove the type assertion.

rosstimothy commented on constants.go:
```diff
@@ -547,17 +547,6 @@ const (
 
-const (
-	// GSuiteIssuerURL is issuer URL used for GSuite provider
-	GSuiteIssuerURL = "https://accounts.google.com"
```
Seems like you could keep this one?

espadolini replied:
Just there in `constants.go`, or private in `lib/auth`?

rosstimothy commented on lib/auth/oidc_google.go:
```diff
@@ -0,0 +1,192 @@
+	return groups, nil
+}
+
+func groupsFromGsuiteCloudidentity(ctx context.Context, email string, clientOptions ...option.ClientOption) ([]string, error) {
```
```suggestion
func groupsFromGsuiteCloudIdentity(ctx context.Context, email string, clientOptions ...option.ClientOption) ([]string, error) {
```

rosstimothy commented on lib/auth/oidc_test.go:
```diff
@@ -176,3 +184,114 @@ func (s *fakeIDP) configurationHandler(w http.ResponseWriter, r *http.Request) {
+			resp.Groups = append(resp.Groups, &directory.Group{Email: groupEmail})
+		}
+
+		b, err := json.Marshal(resp)
+		require.NoError(t, err)
+
+		rw.Write(b)
```
```suggestion
		require.NoError(t, json.NewEncoder(rw).Encode(resp))
```

rosstimothy commented on lib/auth/oidc_test.go:
```diff
@@ -176,3 +184,114 @@ func (s *fakeIDP) configurationHandler(w http.ResponseWriter, r *http.Request) {
+			}
+		}
+
+		b, err := json.Marshal(resp)
+		require.NoError(t, err)
+
+		rw.Write(b)
```
```suggestion
		require.NoError(t, json.NewEncoder(rw).Encode(resp))
```

rosstimothy commented on lib/auth/oidc_google.go:
```diff
@@ -0,0 +1,192 @@
+	}
+
+	var groups []string
+	err = service.Groups.Memberships.SearchTransitiveGroups("groups/-").
+		// the google API docs claim that the query string is a CEL expression
+		// (https://opensource.google/projects/cel) but the call will fail if
+		// you use double quotes instead of single quotes in spite of them being
+		// equivalent according to the CEL specs
+		Query(fmt.Sprintf("member_key_id == '%s' && 'cloudidentity.googleapis.com/groups.discussion_forum' in labels", email)).
```
Where did you get the values for the `SearchTransitiveGroups` and `Query`? Could you add a comment for those of us that aren't as familiar with the cloud identity api as you have become?

rosstimothy reviewed (approved):
Let a few small comments. Looks good otherwise.

espadolini committed:
Address review comments

russjones commented on lib/auth/oidc.go:
```diff
@@ -828,76 +685,9 @@ func (a *Server) getClaims(oidcClient *oidc.Client, connector types.OIDCConnecto
-				return nil, trace.Wrap(err)
-			}
-		}
+	claims, err = addGsuiteClaims(a.closeCtx, connector, claims)
```
I think it's called Google Workspace now, so how about `addGoogleWorkspaceClaims`.

russjones commented on lib/auth/oidc_google.go:
```diff
@@ -0,0 +1,214 @@
+// connector. The credentials in the default case are loaded from the connector
+// with the getGsuiteCredentialsForConnector function.
+func addGsuiteClaims(ctx context.Context, connector types.OIDCConnector, claims jose.Claims, clientOptions ...option.ClientOption) (jose.Claims, error) {
+	// If google_service_account_uri and google_service_account are not set, we
+	// assume that this is a non-GWorkspace OIDC provider using the same
+	// issuer URL as Google Workspace (e.g.
+	// https://developers.google.com/identity/protocols/oauth2/openid-connect).
+	if connector.GetIssuerURL() != gsuiteIssuerURL || (connector.GetGoogleServiceAccountURI() == "" && connector.GetGoogleServiceAccount() == "") {
+		return claims, nil
+	}
```
I'd wrap this into it's own function `isGoogleWorkspace` then you can update `getClaims` to have something like the following:

```go
// If the OIDC provider is configured to be Google Workspace, add in addition claims.
if ok := isGoogleWorkspace(...); ok {
    claims, err = addGoogleWorkspaceClaims(...)
    if err != nil {
        return nil, trace.Wrap(err)
    }
}
return claims, nil
```

russjones commented on lib/auth/oidc_google.go:
```diff
@@ -0,0 +1,214 @@
+	// actual docs for the API call are at
+	// https://cloud.google.com/identity/docs/reference/rest/v1/groups.memberships/searchTransitiveGroups .
+	err = service.Groups.Memberships.SearchTransitiveGroups("groups/-").
+		Query(fmt.Sprintf("member_key_id == '%s' && 'cloudidentity.googleapis.com/groups.discussion_forum' in labels", email)).
```
How does this query language work? Does it have the equivalent of parameterized queries?

I'm wondering if you can do something like SQL injection with it?

espadolini replied:
I tried looking for a "proper" way to do this, but the only library to handle this (https://github.com/google/cel-go) is apparently only for evaluating CEL expressions in text form, not for building them; I don't think it's a problem here, because the `email` parameter we use to format the string is the `email` that came as part of the OIDC claims, so it's something that Google considers to be a valid email address - which I don't believe can include any `'`s.

russjones commented:
@espadolini Just to make sure I understand this correctly, this PR solves two problems.

1. Fetch all domain groups for a user.
2. Fetch transitive group membership for a user.

For (1), the issue is resolved by no longer filtering on a particular domain when we make the [groups.list](https://developers.google.com/admin-sdk/directory/reference/rest/v1/groups/list) query.

For (2), from what I can tell the main issue is that to support getting transitive group membership, we have to call different API endpoints that not all users may have access to due to needing to be on a [Google Workspace Enterprise plan].

Is there any other way to solve (2) or is the Enterprise plan the only way to get transitive group membership for a user?

espadolini commented:
That's correct. The change in default behavior could be mitigated by putting the groups that don't match the domain of the user into a separate claim (`foreign_groups`?) - to maintain the old behavior with the new API we have to do the filtering ourselves, anyway.

I don't think there's a reasonable way to get transitive group memberships without using the intended API - anything ad-hoc that we build ourselves would be too spammy and fragile.

espadolini committed:
Minor refactor and name changes

russjones commented:
@espadolini For (1), let's bump the version of the connector to `OIDCConnectorV3` and in this version drop the domain filter. This way anyone using the old connector will continue to get domain filtering and no unexpected behavior changes but if you upgrade you get all groups.

For (2), can we update the logic to always call `service.Groups.Memberships.SearchTransitiveGroups` first and if it fails, then fallback to `service.Groups.List`? This way we don't have to add an additional field and will still get the same functionality.

espadolini commented:
> @espadolini For (1), let's bump the version of the connector to `OIDCConnectorV3` and in this version drop the domain filter. This way anyone using the old connector will continue to get domain filtering and no unexpected behavior changes but if you upgrade you get all groups.

Ack.

> For (2), can we update the logic to always call `service.Groups.Memberships.SearchTransitiveGroups` first and if it fails, then fallback to `service.Groups.List`? This way we don't have to add an additional field and will still get the same functionality.

Even before calling the actual API we can check if the specified service account is allowed to use the `CloudIdentityGroupsReadonlyScope` scope; if that succeeds we can assume that the system should use the new API, otherwise it should use the old one (and if we also do the check for the `AdminDirectoryGroupReadonlyScope` scope we get a better error message).

I'd be wary of just blindly falling back to the old API on any spurious failure of the new, because if there's roles associated with transitive groups it could lead to a very inconsistent experience for the users - then again, we could just document that for consistent behavior, the service account should only have one of the two scopes enabled. What do you think @russjones?

espadolini committed:
Allow domain filtering, tests now bypass addGoogleWorkspaceClaims

espadolini committed:
Update `OIDCConnectorV2` to `OIDCConnectorV3`

espadolini committed:
Backwards compatibility for OIDCConnector v2

This also removes the extra boolean flag that was added previously.

espadolini committed:
go get google.golang.org/api

go get: upgraded cloud.google.com/go v0.60.0 => v0.100.2
go get: upgraded github.com/golang/snappy v0.0.1 => v0.0.3
go get: upgraded github.com/googleapis/gax-go/v2 v2.0.5 => v2.1.1
go get: upgraded go.opencensus.io v0.22.5 => v0.23.0
go get: upgraded golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d => v0.0.0-20211104180415-d3ed0bb246c8
go get: upgraded google.golang.org/api v0.29.0 => v0.65.0

russjones commented:
@espadolini Please add documentation in another PR on how scopes work now (transitive or regular).

espadolini committed:
Update e-ref

Enterprise builds will break unless gravitational/teleport.e#385
is included.

ibeckermayer commented on lib/auth/oidc_google.go:
```diff
@@ -0,0 +1,256 @@
+				return nil, trace.Wrap(err)
+			}
+		} else {
+			credentials, err := getGoogleWorkspaceCredentials(ctx, connector, directory.AdminDirectoryGroupReadonlyScope)
```
Similar logic to the V2 logic below, consider parameterizing into a function.

espadolini replied:
I had tried refactoring it a bit but this is still the clearest way to write it that I've found, I'm afraid. 😢

ibeckermayer commented on lib/services/oidc.go:
```diff
@@ -102,8 +102,8 @@ func UnmarshalOIDCConnector(bytes []byte, opts ...MarshalOption) (types.OIDCConn
 	switch h.Version {
-	case types.V2:
-		var c types.OIDCConnectorV2
+	case types.V2, types.V3:
```
Perhaps this is standard practice elsewhere in the codebase, in which case ignore this comment, however if not then it may be confusing why we are using an `OIDCConnectorV3` for `case V2` and worthy of a clarifying comment.

espadolini replied:
I have a comment in `(*OIDCConnectorV3).CheckAndSetDefaults()` already, but it's probably useful to restate it here as well.
